README.md
noturno
=======

noturno brought substantial changes in version 1.0.0: the new WebExtensions
API will not let us access the proxy configurations in the browser, as the old
SDK did. This means that now we must get those configurations in noturno,
completely changing the addon core feature. This will also allow us to finally
support other platforms in the future.

Starting on version 2.0.0, noturno dropped support to Firefox < 60 in favor of
supporting Firefox >= 60. This happens due to the deprecation of
`proxy.registerProxyScript` in favor of the new `proxy.onRequest` API event
handler. For noturno ot means that we are now given more control over the proxy
settings, allowing us to easily develop cool features in the future, such as per
proxying requests based on URL filters.

If you have any questions, please open an issue or email me: athoscr AT
fedoraproject.org

Firefox Add-On to switch between proxy settings.
----------------------------------------

https://addons.mozilla.org/en-US/firefox/addon/noturno/

When installed, this Add-On will display a button on the tool bar:

* Whenever the button is gray, proxy settings are set to use no proxy at all.
* If otherwise, the button is green, Firefox will use the proxy set on the **noturno** menu.

### Switching between proxy settings

To switch between no proxy and your proxy settings you may either

* **press the Add-On button on the toolbar and toggle the switch on the menu**
* press **Alt-Shift-X**

### One-click toggle

Starting on version 1.2, noturno supports one click toggle for proxy
configurations. To enable this option,

* right-click noturno button,
* check the "One click toggle" box.

Now you can toggle your proxy configurations just by clicking the button on the
top right of your browser. Note that to change the proxy configurations, like
host or port, you need to deactivate the "One click togle" feature.

### Android support

noturno does not support android for now. The calls bellow are not supported on
Android versions of Firefox, which are used by noturno.

* browser.commands.onCommand.addListener

### Calls to WebExtensions API

Here is a complete list of all calls to the WEbExtensions API. This may be
useful whenever porting noturno to other platforms. 

* browser.action.setIcon
* browser.action.setTitle
* browser.commands.onCommand.addListener
* browser.extension.getURL
* browser.permissions.contains
* browser.permissions.request
* browser.proxy.onError.addListener
* browser.proxy.onRequest.addListener
* browser.runtime.getBackgroundPage
* browser.runtime.onMessage.addListener
* browser.runtime.sendMessage
* browser.storage.local.get
* browser.storage.local.set

### License and Copyright

Copyright 2020 Athos Ribeiro

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)

## [Unreleased]
### Changed
- Use manifest.json v3 format
- Extension now requests host permissions for all sites before first use

### Removed
- Support for Firefox < 101

## [2.1.1] 05-10-2020
### Fixed
- Use safe loads when injecting HTML content

## [2.1.0] 05-10-2020
### Fixed
- Respect browser and system proxy settings when noturno is off

### Changed
- New UI

### Removed
- Removed reset button

## [2.0.0] 05-01-2020
### Added
- Support for Firefox >= 60

### Changed
- Move from deprecated proxy.registerProxyScript to proxy.onRequest

### Removed
- Support for Firefox < 60

## [1.2] 11-16-2017
### Added
- Add option to toggle proxy on/off with one click (activated in the right-click menu).
### Changed
- Proxy DNS queries when using SOCKS.
### Fixed
- Apply proxy changes without the need to toggle off/on when changing configurations while proxy is on.

## [1.1.1] 09-18-2017
### Fixed
- Fix SOCKS5 support.

## [1.1] 09-18-2017
### Added
- Support different protocols like HTTPS and SOCKS.

## [1.0.1] 09-16-2017
### Changed
- Set minimum Firefox version to 55.0. This is the first one to accept WebExtensions proxy API.

## [1.0.0] 09-15-2017
### Added
- Add a menu to the addon.
- Save status (on/off).
- Save last proxy.
- Changed default hotkey to Alt+Shift+X

### Changed
- Port addon to WebExtensions.
- Use addon proxy data instead of Firefox proxy configurations.


test/test-index.js
var index = require("../index");

exports["test checkDefaultType"] = function(assert) {
  assert.equal(5, require("sdk/preferences/service").get("network.proxy.type"), "default config OK");
};

exports["test labels"] = function(assert) {
  assert.equal("Proxy is On - press ctrl+shift+X to turn off", index.proxyOnLabel, "On label OK");
  assert.equal("Proxy is Off - press ctrl+shift+X to turn on", index.proxyOffLabel, "Off label OK");
};

exports["test toggleProxyConfig"] = function(assert) {
  index.toggleProxyConfig();
  assert.equal(1, require("sdk/preferences/service").get("network.proxy.type"), "toggle works");
};

require("sdk/test").run(exports);


pac.js
/*
 * As described in
 * https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/proxy
 * and
 * https://bugzilla.mozilla.org/show_bug.cgi?id=1355198
 * require a parameter to avoid a "Invalid Proxy Rule: DIRECT" error
*/
const noProxy = "DIRECT stub_param";

var proxy = noProxy;
var proxyOn = false;

browser.runtime.sendMessage("PAC script initialized");

browser.runtime.onMessage.addListener((message) => {
  proxyOn = message.enabled;
  if(proxyOn && message.host && message.port && message.protocol) {
    browser.runtime.sendMessage("Setting proxy to " + message.protocol + " " + message.host + ":" + message.port);
    proxy = [
      {
        type: message.protocol,
        host: message.host,
        port: message.port,
      }
    ]
    if(proxy[0].type == "SOCKS" || proxy[0].type == "SOCKS4") {
      proxy[0].proxyDNS = true
    }
  } else if(proxyOn) {
    browser.runtime.sendMessage("Host, Port or Protocol missing, setting proxy to DIRECT");
    proxy = noProxy;
  } else {
      browser.runtime.sendMessage("Proxy off");
      proxy = noProxy;
  }
});

function FindProxyForURL(url, host) {
  return proxy;
}



menu/menu.js
var toggle_switch = document.getElementById('toggle-switch');
var icon = document.getElementById('icon');
var host = document.getElementById('host');
var port = document.getElementById('port');
var type = document.getElementById('type');
var form = document.getElementById('settings-form');

var typeErr = document.getElementById("error-type")
var hostErr = document.getElementById("error-host")
var portErr = document.getElementById("error-port")

var displayTabButton = document.getElementById("display-tab-button")
var settingsTabButton = document.getElementById("settings-tab-button")

var tabButtons = document.querySelectorAll('.panel-section-tabs-button');
var tabContents = document.querySelectorAll('.tab-content');

displayTabButton.onclick = function() {
  changeTab(event, "display-tab")
}

settingsTabButton.onclick = function() {
  changeTab(event, "settings-tab")
}

function changeTab(evt, tabId) {
  console.debug("Changing to tab " + tabId);

  for(i = 0; i < tabButtons.length; i++) {
    if(tabButtons[i].id == evt.currentTarget.id) {
      tabButtons[i].setAttribute("class", "panel-section-tabs-button selected")
    } else {
      tabButtons[i].setAttribute("class", "panel-section-tabs-button")
    }
  }

  for(i = 0; i < tabContents.length; i++) {
    tabContents[i].style.display = "none";
  }

  tab = document.getElementById(tabId)
  tab.style.display = "block";
}

function setInfoPage(type, host, port) {
  if(!host){
    document.getElementById("host-info").textContent = "NO PROXY";
    return
  }
  document.getElementById("type-info").textContent = type.toUpperCase();
  document.getElementById("host-info").textContent = host;
  document.getElementById("port-info").textContent = port;
}

function init(page) {
  page.getProxyConfig();
  host.value = page.proxy.host;
  port.value = page.proxy.port;
  type.value = page.proxy.type;
  toggle_switch.checked = page.proxy.enabled;
  setInfoPage(page.proxy.type, page.proxy.host, page.proxy.port)
}

function toggle(page) {
  page.toggleProxyConfig();
}

function save(page) {
  page.proxy.host = host.value;
  page.proxy.port = port.value;
  page.proxy.type = type.value;
  page.saveProxyConfig();
}

function onError(error) {
  console.log(`Error: ${error}`);
}

var get_background = browser.runtime.getBackgroundPage();

/* initialize values */
get_background.then(init, onError);

host.oninput = function() {
  hostErr.setAttribute("class", "text-error")
  host.setCustomValidity("");
}

port.oninput = function() {
  portErr.setAttribute("class", "text-error")
  port.setCustomValidity("");
}

type.onchange = function() {
  typeErr.setAttribute("class", "text-error")
  type.setCustomValidity("");
}

form.onsubmit = function() {
  var valid = true
  if(!port.value) {
    var errMsg = "Port should not be blank"
    port.setCustomValidity(errMsg);
    portErr.textContent = errMsg
    portErr.setAttribute("class", "text-error display-block")
    valid = false
  }
  if(!host.value) {
    var errMsg = "Host should not be blank"
    host.setCustomValidity(errMsg);
    hostErr.textContent = errMsg
    hostErr.setAttribute("class", "text-error display-block")
    valid = false
  }
  if(!type.value) {
    var errMsg = "Select a protocol"
    type.setCustomValidity(errMsg);
    typeErr.textContent = errMsg
    typeErr.setAttribute("class", "text-error display-block")
    valid = false
  }
  if(!valid) {
    return false
  }
  get_background.then(save, onError);
  return true
}

toggle_switch.onclick = function() {
  get_background.then(toggle, onError);
}

// Toggle the switch when hotkey is activated
function handleMessage(message, sender) {
  if(message == "toggle-switch") {
    toggle_switch.checked = !(toggle_switch.checked);
  }
}

browser.runtime.onMessage.addListener(handleMessage);


background.js
// Log errors from proxy script
browser.proxy.onError.addListener(error => {
  console.error(`Proxy error: ${error.message}`);
});

var proxy = {
  enabled: false,
  type: '',
  host: '',
  port: '',
  filters: {urls: ["<all_urls>"]}
}

function proxyListener(requestDetails) {
  proxyInfo = proxy
  console.debug('proxying request: ', requestDetails);
  return proxyInfo
}

const proxyOnIcon = {
  "16": "data/on_16.png",
  "32": "data/on_32.png",
  "64": "data/on_64.png"
};

const proxyOffIcon = {
  "16": "data/off_16.png",
  "32": "data/off_32.png",
  "64": "data/off_64.png"
};

const proxyPermission = {
  origins: ["<all_urls>"]
};

const proxyOnLabel = "Proxy is On - press Alt+Shift+X to turn off";
const proxyOffLabel = "Proxy is Off - press Alt+Shift+X to turn on";

var oneClickToggle = false

function init() {
  browser.storage.local.get()
    .then((storedSettings) => {
      requiredKeys = ["enabled", "port", "host", "type"]
      hasRequiredKeys = requiredKeys.every(key => Object.keys(storedSettings).includes(key))
      hasNoRequiredKeys = requiredKeys.every(key => !(Object.keys(storedSettings).includes(key)))
      if(hasRequiredKeys) {
        proxy.enabled = storedSettings.enabled;
        proxy.host = storedSettings.host;
        proxy.port = storedSettings.port;
        proxy.type = storedSettings.type;
        if(proxy.enabled) {
          browser.action.setIcon({path: proxyOnIcon});
          browser.action.setTitle({title: proxyOnLabel});
          browser.proxy.onRequest.addListener(proxyListener, proxy.filters);
        }
      } else if(hasNoRequiredKeys) {
        saveProxyConfig();
      }
      if("oneClickToggle" in storedSettings) {
        oneClickToggle = storedSettings.oneClickToggle
        toggleOneClickToggle(oneClickToggle)
      }
    });
}

function getProxyConfig() {
  browser.storage.local.get()
    .then((storedSettings) => {
      proxy.enabled = storedSettings.enabled;
      proxy.host = storedSettings.host;
      proxy.port = storedSettings.port;
      proxy.type = storedSettings.type;
    });
}

function saveProxyConfig() {
    browser.storage.local.set(proxy);
}

async function requestPermission() {
  const response = await browser.permissions.request(proxyPermission);
  if (response) {
    console.log("Permission was granted");
    browser.action.setPopup({popup: "menu/menu.html"})
  } else {
    console.log("Permission was denied");
  }
}

function toggleProxyConfig() {
  proxy.enabled = !proxy.enabled;
  if(proxy.enabled) {
    console.debug("Proxy has been turned ON")
    browser.action.setIcon({path: proxyOnIcon});
    browser.action.setTitle({title: proxyOnLabel});
    browser.proxy.onRequest.addListener(proxyListener, proxy.filters);
  }
  else {
    console.debug("Proxy has been turned OFF")
    browser.action.setIcon({path: proxyOffIcon});
    browser.action.setTitle({title: proxyOffLabel});
    browser.proxy.onRequest.removeListener(proxyListener);
  }
  saveProxyConfig();
}

function toggleOneClickToggle(checked) {
  if(checked){
    browser.action.setPopup({popup: ""})
    browser.action.onClicked.addListener(toggleProxyConfig)
  } else {
    browser.action.setPopup({popup: "menu/menu.html"})
  }
  oneClickToggle = checked
  browser.storage.local.set({oneClickToggle: checked});
}

init();

async function checkPermission() {
  const has_permission = await browser.permissions.contains(proxyPermission);
  if (!has_permission) {
      browser.action.setPopup({popup: ""})
      browser.action.onClicked.addListener(requestPermission)
  }
}
checkPermission();

browser.commands.onCommand.addListener(function(command) {
  if (command == "toggle-status") {
    // This message makes the popup toggle the switch when open
    browser.runtime.sendMessage('toggle-switch');
    toggleProxyConfig();
  }
});

function createMenu(details) {
  console.log(details.reason);
  browser.menus.create({
    id: "oneClickToggle",
    type: "checkbox",
    title: "One click toggle",
    contexts: ["browser_action"],
    checked: oneClickToggle
  });
}

browser.runtime.onInstalled.addListener(createMenu);

browser.menus.onClicked.addListener(function(info, tab) {
  if (info.menuItemId == "oneClickToggle") {
    toggleOneClickToggle(info.checked)
  }
});


index.js
var self = require('sdk/self');

var proxyOnIcon = {
  "16": "./on_16.png",
  "32": "./on_32.png",
  "64": "./on_64.png"
};

var proxyOffIcon = {
  "16": "./off_16.png",
  "32": "./off_32.png",
  "64": "./off_64.png"
};

var proxyOnLabel = "Proxy is On - press ctrl+shift+X to turn off";
var proxyOffLabel = "Proxy is Off - press ctrl+shift+X to turn on";

var initialIcon = proxyOnIcon;
var initialLabel = proxyOnLabel;

var service = require("sdk/preferences/service");

// possible values for network.proxy.type
var proxyNone = 0; // No proxy
var proxyManual = 1; // Manual proxy configuration
var proxyAuto = 4; // Auto-detect proxy settings for this network
var proxySystem = 5; // Use system proxy settings (Firefox default)

var proxyOff = proxySystem;
var proxyOn = proxyManual;

// TODO: save user proxyOff configuration in case she closes the browser
//  note that not doing so will possibly change user's proxyOff configuration
//  whenever she re-opens Firefox
switch(service.get("network.proxy.type")) {
  case proxyNone:
    proxyOff = proxyNone;
    break;
  case proxyAuto:
    proxyOff = proxyAuto;
    break;
}

if(service.get("network.proxy.type") == proxyOff) {
  initialIcon = proxyOffIcon;
  initialLabel = proxyOffLabel;
}

var { ActionButton } = require("sdk/ui/button/action");
var button = ActionButton({
  id: "noturno-state",
  label: initialLabel,
  icon: initialIcon,
  onClick: toggleProxyConfig
});

var { Hotkey } = require("sdk/hotkeys");
var showHotKey = Hotkey({
  combo: "accel-shift-x",
  onPress: toggleProxyConfig
});

function toggleProxyConfig() {
  if(service.get("network.proxy.type") == proxyOff) {
    service.set("network.proxy.type", proxyOn);
    button.icon = proxyOnIcon;
    button.label = proxyOnLabel;
  }
  else {
    service.set("network.proxy.type", proxyOff);
    button.icon = proxyOffIcon;
    button.label = proxyOffLabel;
  }
}

module.exports.proxyOnLabel = proxyOnLabel
module.exports.proxyOffLabel = proxyOffLabel
module.exports.toggleProxyConfig = toggleProxyConfig

